### PR TITLE
Allow mock connector to return data

### PR DESCRIPTION
**Pull request by kokosing**
Allow mock connector to return data

Mock connector just pretends that it supports INSERT, DELETE and UPDATE.
It does not modify any data. Added tests for that.

### CODE REVIEW

**Comment by losipiuk** [comment]
Commit title is somewhat misleading. Maybe just `Allow mock connector to produce data`. Manipulating is not really supported (as you write later in PR desc).

**Review by MiguelWeezardo — APPROVED** [review]

File: testing/trino-tests/src/test/java/io/trino/tests/TestMockConnector.java [hunk 1]
```diff
@@ -102,4 +116,45 @@ public void testDropMaterializedView()
...
+                "SELECT nationkey, regionkey FROM nation");
+        assertQuery(
+                "SELECT regionkey FROM mock.default.nation",
+                "SELECT regionkey FROM nation");
+    }
+
+    @Test
+    public void testInsert()
+    {
+        assertUpdate("INSERT INTO mock.default.nation VALUES (101, 'POLAND', 0, 'No comment')", 1);
```
Comment (MiguelWeezardo) [inline testing/trino-tests/src/test/java/io/trino/tests/TestMockConnector.java:140]: nit: Assert that no actual INSERT took place

**Comment by kokosing** [comment]
@losipiuk @MiguelWeezardo AC, PTAL

File: core/trino-main/src/test/java/io/trino/connector/MockConnector.java [hunk 2]
```diff
@@ -588,23 +594,63 @@ public ConnectorPageSink createPageSink(ConnectorTransactionHandle transactionHa
...
+            List<List<?>> records = data.apply(tableName).stream()
+                    .map(record -> {
+                        ImmutableList.Builder<Object> projectedRow = ImmutableList.builder();
+                        for (int columnIndex = 0; columnIndex < columnNames.size(); columnIndex++) {
+                            if (projection.contains(columnNames.get(columnIndex))) {
+                                projectedRow.add(record.get(columnIndex));
+                            }
+                        }
+                        // produce artificial id column used by DELETE and UPDATE
+                        projectedRow.add(0);
```
Comment (losipiuk) [inline core/trino-main/src/test/java/io/trino/connector/MockConnector.java:625]: I do not understand how this one is used. Seems not.
Also don't you need to override `getUpdateRowIdColumnHandle` and `getDeleteRowIdColumnHandle` in `MockConnectorMetadata` for `DELETE`/`UPDATE` flow to work?

File: testing/trino-tests/src/test/java/io/trino/tests/TestMockConnector.java (hunk 1 above)
Comment (losipiuk) [reply testing/trino-tests/src/test/java/io/trino/tests/TestMockConnector.java:141]: +1
same for delete and update

Comment (losipiuk) [reply testing/trino-tests/src/test/java/io/trino/tests/TestMockConnector.java:141]: And comment why we do not expect INSERT/DELETE/UPDATE to actually happen

**Review by losipiuk — APPROVED** [review]
looks good generally.

File: core/trino-main/src/test/java/io/trino/connector/MockConnector.java (hunk 2 above)
Comment (kokosing) [reply core/trino-main/src/test/java/io/trino/connector/MockConnector.java:625]: > Also don't you need to override getUpdateRowIdColumnHandle and getDeleteRowIdColumnHandle in MockConnectorMetadata for DELETE/UPDATE flow to work?

This was already added to the repo (I guess by @Praveen2112 ), in the meantime when my code was in the review. I added here only tests for UPDATE and DELETE.

> I do not understand how this one is used. Seems not.

Will try to remove it. I recall I added it for a reason, but I do see that comment is not the best one.

Comment (kokosing) [reply core/trino-main/src/test/java/io/trino/connector/MockConnector.java:625]: Without this it fails:
```
java.lang.RuntimeException: java.lang.ArrayIndexOutOfBoundsException

	at io.trino.testing.AbstractTestingTrinoClient.execute(AbstractTestingTrinoClient.java:120)
	at io.trino.testing.DistributedQueryRunner.execute(DistributedQueryRunner.java:476)
	at io.trino.testing.QueryAssertions.assertUpdate(QueryAssertions.java:71)
	at io.trino.testing.AbstractTestQueryFramework.assertUpdate(AbstractTestQueryFramework.java:238)
	at io.trino.testing.AbstractTestQueryFramework.assertUpdate(AbstractTestQueryFramework.java:233)
	at io.trino.tests.TestMockConnector.testUpdate(TestMockConnector.java:161)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at org.testng.TestRunner.privateRun(TestRunner.java:756)
	at org.testng.TestRunner.run(TestRunner.java:610)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:387)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:382)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1293)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1218)
	at org.testng.TestNG.runSuites(TestNG.java:1133)
	at org.testng.TestNG.run(TestNG.java:1104)
	at com.intellij.rt.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:66)
	at com.intellij.rt.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:109)
	Suppressed: java.lang.Exception: SQL: UPDATE mock.default.nation SET name = 'ALGERIA'
		at io.trino.testing.DistributedQueryRunner.execute(DistributedQueryRunner.java:479)
		... 28 more
Caused by: java.lang.ArrayIndexOutOfBoundsException

```

Simalarly for DELETE.

**Comment by kokosing** [comment]
CI hit: https://github.com/trinodb/trino/issues/8911

File: core/trino-main/src/test/java/io/trino/connector/MockConnector.java (hunk 2 above)
Comment (losipiuk) [reply core/trino-main/src/test/java/io/trino/connector/MockConnector.java:625]: Actually I think this is not correct. IIUC in returned PageSource the blocks order should match column handles ordering from `columns` argument.
List of column handles passed here can be in arbitrary order, also the update/delete row_id column does not have to be last one.

Comment (kokosing) [reply core/trino-main/src/test/java/io/trino/connector/MockConnector.java:625]: Maybe I haven't tried enough. But I was not able to see update/delete row_id column in columns argument. Other than that column the projection is supported. See `testDataGeneration` where we test different projections.

Comment (losipiuk) [reply core/trino-main/src/test/java/io/trino/connector/MockConnector.java:625]: image

Comment (kokosing) [reply core/trino-main/src/test/java/io/trino/connector/MockConnector.java:625]: https://github.com/trinodb/trino/pull/8914 Thanks!

**Pull request merged**